### PR TITLE
Patterns: tracking of pattern permalink activity

### PR DESCRIPTION
**Pull request by nightnei**
Related to https://github.com/Automattic/dotcom-forge/issues/6460

## Proposed Changes
1) With this PR we are adding property `name` to `calypso_pattern_library_view` which has pattern name as value if the direct link to pattern was opened.
2) Added tracking event to "copy permalink".
3) Introduced `useRecordPatternsEvent` hook. Reason - we have a lot of copy-paste, so with this hook we can avoid copy-paste (AFAIS it's `category` and `type`); additionally - I see that we always track different properties, so I would consider the list of default props one time and we will not miss them: if we have always `type` so maybe would be good to always additionally include `view`; maybe also would be nice to always include `is_logged_in` and `user_is_dev_account`. @wongasy WDYT about my idea?

## Testing Instructions
1.1) Open `/patterns/events#pattern-14736`
1.2) **Wait a bit**
1.3) Assert that `calypso_pattern_library_view` is tracked w/ property as `name: events-list`
1.4) Open `/patterns/events`
1.5) Assert that `calypso_pattern_library_view` is tracked w/o `name` property

2.1) Click on any pattern permalink to copy it
2.2) Assert that `calypso_pattern_permalink_copy` is tracked with `category: events, type: pattern, name: <PATTERN_NAME>`

### CODE REVIEW

File: client/my-sites/patterns/components/pattern-preview/index.tsx
Comment (wongasy) [inline client/my-sites/patterns/components/pattern-preview/index.tsx]:
```suggestion
							recordPatternsEvent( 'calypso_pattern_library_permalink_copy', {
```
Let's stick with the same prefix for all pattern library events.

File: client/my-sites/patterns/components/pattern-library/index.tsx
Comment (wongasy) [inline client/my-sites/patterns/components/pattern-library/index.tsx]: I would suggest relying on `isFetchingPatterns` to determine the availability of the `patterns` array, and we can eliminate indentation as follows:

```suggestion
	if ( ! isFetchingPatterns && patternPermalinkId ) {
		const pattern = patterns.find( ( pattern ) => pattern.ID === patternPermalinkId );
		patternPermalinkName = pattern ? pattern.name : undefined;
	}
```

File: client/my-sites/patterns/hooks/use-record-patterns-event.tsx
Comment (wongasy) [inline client/my-sites/patterns/hooks/use-record-patterns-event.tsx]:
```suggestion
				category: category || undefined,
```

Category may be empty in the search result page. This avoids reporting an empty `category` property.

**Review by wongasy — COMMENTED** [review]
>we have a lot of copy-paste, so with this hook we can avoid copy-paste (AFAIS it's category and type); additionally - I see that we always track different properties, so I would consider the list of default props one time and we will not miss them

I don't mind introducing a hook to handle common properties. That said, I don't think it's of priority to refactor all existing calls to use this hook. We also need to be careful of what properties should/should not be included for different events. For example `type` and `view` should not be tracked in the search results page and the main page (`/patterns`). The expected properties for each event are listed out here pcYYuD-1hV-p2 (some requirements might be out of date, but the post should provide a good reference). 

In _some_ instances, it isn't necessarily a bad thing to copy-paste, it gives us better control of which exact properties to supply for each event, not having to worry about what the default behaviour is of the hook.

In the interest of wrapping up this particular issue https://github.com/Automattic/dotcom-forge/issues/6460, I'm fine with moving forward with what you have here. Future refactoring can be discussed in a different conversation.

File: client/my-sites/patterns/components/pattern-library/index.tsx
Comment (stephanethomas) [reply client/my-sites/patterns/components/pattern-library/index.tsx]: I think we could even simplify the two last lines from Angela's suggestion like so:

```javascript
patterns.find( pattern => pattern.ID === patternPermalinkId )?.name;
```

File: client/my-sites/patterns/index.web.tsx
Comment (stephanethomas) [inline client/my-sites/patterns/index.web.tsx]: We should probably move that function to [this file](https://github.com/Automattic/wp-calypso/blob/7059acc72c7c1527d47b17959ad52a48f41694a9/client/my-sites/patterns/paths.ts).

File: client/my-sites/patterns/components/page-view-tracker.tsx
Comment (stephanethomas) [inline client/my-sites/patterns/components/page-view-tracker.tsx:62]: Should we record that property even if it is empty? I remember we discussed ditching `search_term` if it is empty but I'm surprised to see that we don't do that here.

**Review by stephanethomas — APPROVED** [review]
This works as expected. I've left minor comments but I'll let you decide if it's worth addressing them. This is good to go as long as Angela's feedback is also addressed 🙂.

File: client/my-sites/patterns/index.web.tsx
Comment (fredrikekelund) [reply client/my-sites/patterns/index.web.tsx]: Alternatively to `client/my-sites/patterns/lib/get-pattern-permalink.ts`

Comment (nightnei) [reply client/my-sites/patterns/index.web.tsx]: >Alternatively to client/my-sites/patterns/lib/get-pattern-permalink.ts

Yeah, I think it's better place, thanks for ideas

File: client/my-sites/patterns/components/page-view-tracker.tsx
Comment (nightnei) [reply client/my-sites/patterns/components/page-view-tracker.tsx:62]: IMO it doesn't make sense to have this empty property on, for example, home page, since it's not possible to have permalink pattern there or any other type of `name`

File: client/my-sites/patterns/hooks/use-record-patterns-event.tsx
Comment (nightnei) [reply client/my-sites/patterns/hooks/use-record-patterns-event.tsx]: Hm, maybe I am missing something but it seems it doesn't make any sense, unless we provide `''` (empty string).
Am I wrong?

Comment (nightnei) [reply client/my-sites/patterns/hooks/use-record-patterns-event.tsx]: By "empty", do you mean that - to avoid tracking "empty string" (`""`) we should set `undefined`?

File: client/my-sites/patterns/components/page-view-tracker.tsx
Comment (nightnei) [reply client/my-sites/patterns/components/page-view-tracker.tsx:62]: Also, just to confirm - by `empty ` did you mean to provide "empty string"? AFAIU it's only one way to record property as "empty', right?

File: client/my-sites/patterns/hooks/use-record-patterns-event.tsx
Comment (wongasy) [reply client/my-sites/patterns/hooks/use-record-patterns-event.tsx]: >avoid tracking "empty string" ("") we should set undefined

Correct!

File: client/my-sites/patterns/components/page-view-tracker.tsx
Comment (wongasy) [inline client/my-sites/patterns/components/page-view-tracker.tsx:62]:
```suggestion
				name: patternPermalinkName || undefined,
```

This prevents us from sending an empty value to the `name` property. i.e. `name` will only be included in the tracking call if the name exists.

Comment (nightnei) [reply client/my-sites/patterns/components/page-view-tracker.tsx:62]: We don't need it, since I implemented `patternPermalinkName` as `undefiend` if it's not permalink

Comment (stephanethomas) [reply client/my-sites/patterns/components/page-view-tracker.tsx:62]: Yes, I meant an empty string as `''`.

File: client/my-sites/patterns/lib/get-pattern-permalink.ts
Comment (wongasy) [inline client/my-sites/patterns/lib/get-pattern-permalink.ts]: Nit. I would probably move this to its own file, since the current setup is such that the file name matches the function name. But I see that there's been a few suggestions on where to place this function in previous reviews, which is why we landed the function here now. I'll leave it up to you to decide.

File: client/my-sites/patterns/lib/get-pattern-permalink.ts
Comment (stephanethomas) [inline client/my-sites/patterns/lib/get-pattern-permalink.ts]: I'm on the fence - I can understand why we would put that here given it relates to pattern permalinks. However, the file is named after the other function so it's a bit confusing to see it also exports that new function. Said differently: should we rename this file to something more generic, or move that new function to it's own file?

I think we should also revisit whether we should keep `paths.ts` if some functions that relate to paths/urls are in a dedicated file while some others are located in that particular file.

File: client/my-sites/patterns/hooks/use-record-patterns-event.tsx
Comment (wongasy) [inline client/my-sites/patterns/hooks/use-record-patterns-event.tsx:15]:
```suggestion
				type: getTracksPatternType( patternTypeFilter ),
				view: isGridView ? 'grid' : 'list',
```
I just came to realize that, we only use this hook in pages other than the main page `/patterns`. So it's fine to default `view` property to `list` (i.e. default to `list` in search result).

**Review by wongasy — APPROVED** [review]

File: client/my-sites/patterns/hooks/use-record-patterns-event.tsx
Comment (wongasy) [inline client/my-sites/patterns/hooks/use-record-patterns-event.tsx]:
```suggestion
	const { category, isGridView, patternTypeFilter } = usePatternsContext();
```

**Review by wongasy — APPROVED** [review]
I've tested the changes locally applying this suggestion https://github.com/Automattic/wp-calypso/pull/89270/files#r1555732426 Tracking works as expected!

#### Event fired when copying permalink in list view

Screen Shot 2024-04-08 at 8 30 59 AM

#### Event fired when copying permalink in grid view

Screen Shot 2024-04-08 at 8 31 05 AM

#### `name` property included in `calypso_pattern_library_view` when arriving via permalink

Screen Shot 2024-04-08 at 8 28 37 AM

File: client/my-sites/patterns/hooks/use-record-patterns-event.tsx
Comment (nightnei) [reply client/my-sites/patterns/hooks/use-record-patterns-event.tsx:15]: Is it unintentionally deleted `type: getTracksPatternType( patternTypeFilter ),`? Shoudl we keep both `type` and `view`?

Comment (wongasy) [reply client/my-sites/patterns/hooks/use-record-patterns-event.tsx:15]: No, I didn't mean to delete `type`. Sorry!

Suggestion adjusted!

**Pull request merged**
